### PR TITLE
[OM-88273]: GitOps Configuration Overrides via CustomResource

### PR DESCRIPTION
**Pull request by amd-ibm**
## Intent
Integrate the [new GitOps CustomResource](https://github.com/turbonomic/turbo-gitops) into kubeturbo

## Background
Recently, support was added for applications controlled via GitOps (ArgoCD). The configuration for this functionality is provided as runtime parameters, and limits users to a single git account (only one set of credentials can be provided), a single commit mode type for all applications, etc. The changes here, integrate a new GitOps CustomResource into kubeturbo that allows users to override these default configurations, such as overriding the commit mode or providing a separate set of credentials.

NOTE: The GitOps CustomResources are namespace-scoped, so overrides will only apply to apps within the same namespace as  the resource is deployed.

## Testing
Spun up multiple GitOps apps in appliance, and tested kubeturbo locally against remote appliance. Tests included the following configuration scenarios:
* tested custom resource with selector
* tested custom resource with whitelist
* tested custom resource with both selector and whitelist
* tested custom resource without either selector or whitelist
* tested custom resource with credentials

Additionally, added and updated unit tests.

### CODE REVIEW

**Review by pankajbatra1 — APPROVED** [review]

File: pkg/cluster/cluster_info_scraper.go [hunk 1]
```diff
@@ -66,17 +69,21 @@ type ClusterScraper struct {
 	DynamicClient           dynamic.Interface
 	ControllerRuntimeClient runtimeclient.Client
 	cache                   turbostore.ITurboCache
+	GitOpsConfigCache       *map[string][]*gitopsv1alpha1.Configuration
```
Comment (irfanurrehman) [inline pkg/cluster/cluster_info_scraper.go:72]: maps in go are already pointers and can be exactly used as pointers. You would almost never need a pointer to a map.

File: pkg/action/executor/k8s_controller_updater.go [hunk 2]
```diff
@@ -74,11 +74,13 @@ func newK8sControllerUpdater(clusterScraper *cluster.ClusterScraper, ormClient *
...
-			managerApp:   managerApp,
-			gitConfig:    gitConfig,
-			k8sClusterId: clusterId,
+			name:                  kind,
+			ormClient:             ormClient,
+			managerApp:            managerApp,
+			gitConfig:             gitConfig,
+			k8sClusterId:          clusterId,
+			gitOpsConfigCache:     *clusterScraper.GitOpsConfigCache,
+			gitOpsConfigCacheLock: &clusterScraper.GitOpsConfigCacheLock,
```
Comment (irfanurrehman) [inline pkg/action/executor/k8s_controller_updater.go:83]: I dont generally define a pointer variable until there is a need for one. I don't understand the need here.

File: pkg/discovery/processor/cluster_processor.go [hunk 3]
```diff
@@ -229,6 +229,14 @@ func (p *ClusterProcessor) DiscoverCluster() (*repository.ClusterSummary, error)
 	// Discover Turbo Policies
 	NewTurboPolicyProcessor(p.clusterInfoScraper, kubeCluster).ProcessTurboPolicies()
 
+	// Discover GitOps configuration overrides
+	NewGitOpsConfigProcessor(p.clusterInfoScraper, kubeCluster).ProcessGitOpsConfigs()
+
+	// Update the GitOps configuration override cache
+	if clusterScraper, ok := p.clusterInfoScraper.(*cluster.ClusterScraper); ok {
+		clusterScraper.UpdateGitOpsConfigCache(kubeCluster.GitOpsConfigurations)
```
Comment (irfanurrehman) [inline pkg/discovery/processor/cluster_processor.go:237]: This is an odd usage. We are sort of duplicating the cache data in memory. Effectively storing the copy in `kubeCluster` and then in `clusterScraper.GitopsConfigCache`. We need only one copy. I would simply call `GetAllGitOpsConfigurations()` here and pass the data on. We don't need the copy in `kubeCluster`. Or do we, I might be missing something.. :-)

File: pkg/discovery/processor/cluster_processor.go [hunk 4]
```diff
@@ -229,6 +229,14 @@ func (p *ClusterProcessor) DiscoverCluster() (*repository.ClusterSummary, error)
 	// Discover Turbo Policies
 	NewTurboPolicyProcessor(p.clusterInfoScraper, kubeCluster).ProcessTurboPolicies()
 
+	// Discover GitOps configuration overrides
+	NewGitOpsConfigProcessor(p.clusterInfoScraper, kubeCluster).ProcessGitOpsConfigs()
+
+	// Update the GitOps configuration override cache
+	if clusterScraper, ok := p.clusterInfoScraper.(*cluster.ClusterScraper); ok {
```
Comment (irfanurrehman) [inline pkg/discovery/processor/cluster_processor.go:236]: Why do we need the type check here?  We already defined and implemented the `GetAllGitOpsConfigurations()` interface.

File: pkg/discovery/processor/gitops_config_processor.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package processor
```
Comment (irfanurrehman) [inline pkg/discovery/processor/gitops_config_processor.go:1]: I am not sure if we really need this processor.

**Review by irfanurrehman — CHANGES_REQUESTED** [review]
Thanks @amd-ibm for doing this.
Few changes suggested.

File: pkg/action/executor/k8s_controller.go [hunk 6]
```diff
@@ -251,3 +258,76 @@ func (c *parentController) shouldSkipOperator(controller *unstructured.Unstructu
...
+}
+
+// Returns a flag indicating if there exists a GitOps configuration override for the supplied application
+func isGitOpsConfigOverridden(config *gitopsv1alpha1.Configuration, objName string) bool {
+	return isAppSelectorMatch(config.Selector, objName) || isAppInWhitelist(config.Whitelist, objName)
+}
+
+// Returns the credentials to be used for a GitOps operation.
+func getGitOpsCredentials(overrideConfig *gitopsv1alpha1.Configuration, defaultConfig gitops.GitConfig) (string, string, string, string) {
+	secretNamespace := defaultConfig.GitSecretNamespace
```
[the remark this answers is not in the document]
Comment (amd-ibm) [reply pkg/action/executor/k8s_controller.go]: Updated

File: pkg/action/executor/k8s_controller.go [hunk 7]
```diff
@@ -251,3 +258,76 @@ func (c *parentController) shouldSkipOperator(controller *unstructured.Unstructu
...
+		glog.V(4).Infof("Overriding GitOps credentials to use username [%v] instead of the default.", username)
+	}
+	return email, username, secretName, secretNamespace
+}
+
+// Returns the GitOps configuration for the supplied application. If an override exists, it will return it. Otherwise,
+// it will return the default configuration supplied that was supplied as a runtime parameter.
+func ( c *parentController) GetGitOpsConfig(obj *unstructured.Unstructured) gitops.GitConfig {
+	appName := c.managerApp.Name
+	glog.V(2).Infof("Checking for GitOps configuration override for %v...", appName)
```
[the remark this answers is not in the document]
Comment (amd-ibm) [reply pkg/action/executor/k8s_controller.go]: Updated the log level

File: pkg/cluster/cluster_info_scraper.go (hunk 1 above)
Comment (amd-ibm) [reply pkg/cluster/cluster_info_scraper.go]: Removed the pointer.

File: pkg/action/executor/k8s_controller_updater.go (hunk 2 above)
Comment (amd-ibm) [reply pkg/action/executor/k8s_controller_updater.go:83]: Are you referring to the lock here?

File: pkg/discovery/processor/cluster_processor.go (hunk 3 above)
Comment (amd-ibm) [reply pkg/discovery/processor/cluster_processor.go]: Updated to call `UpdateGitOpsConfigCache` which in turn calls `GetAllGitOpsConfigurations`.

File: pkg/discovery/processor/cluster_processor.go (hunk 4 above)
Comment (amd-ibm) [reply pkg/discovery/processor/cluster_processor.go]: Removed this code.

File: pkg/discovery/processor/gitops_config_processor.go (hunk 5 above)
Comment (amd-ibm) [reply pkg/discovery/processor/gitops_config_processor.go]: Removed the processor.

File: pkg/action/executor/k8s_controller_updater.go (hunk 2 above)
Comment (irfanurrehman) [reply pkg/action/executor/k8s_controller_updater.go:83]: Ignore this plz.. :-)

**Review by irfanurrehman — APPROVED** [review]
Thanks @amd-ibm for the updates.
I am not sure, why I see multiple individual commits with the same commit msg. Please ensure to squash the commits before merge.

**Pull request merged**